topics/data-science/tutorials/python-loops/tutorial.md

```diff
@@ -0,0 +1,528 @@
+
+```
+for sequence in sequences:
+    print()
+for patient in clinic_patients:
+    print()
+for nucleotide in dna_sequence:
+    print()
+```
+
+## Range
+
+You can use `range` to iterate over a sequence of numbers. This is a built in function (check `help(range)`!) so it's always available even if you don't `import` anything. The range produced is non-inclusive: `range(N)` is the numbers `0` to `N-1`, so the result will be exactly the length you requested.
+
+```python
+for number in range(3):
+    print(number)
+```
+
+However in python `range` is a special type of iterable: none of the numbers are created until we need them.
+
+```python
+print(range(5))
+print(range(-3, 8)[0:4])
+```
+
+The easiest way to see what numbers are actually in there is to convert it to a `list`:
+
+```
+print(list(range(5)))
+print(list(range(-3, 8)))
+print(list(range(0, 10, 2)))
+```
+
+## Accumulation
+
+In programming you'll often want to accumulate some values: counting things (or "accumulating"). The pattern consists of creating a variable to store your result, running a loop over some data, and in that loop, adding to the variable for your result.
+
+```python
+# Sum the first 10 integers.
+total = 0
+for number in range(1, 11):
+    total = total + (number)
+print(total)
+```
+
+But how did we get that result? We can add some "debugging" lines to the above code to figure out how we got to that result. Try adding the following line in the above loop
+
+```
+print(f'Currently {number}, our total is {total}')
+```
+
+You can add it before you update `total`, after it, or both! Compare the outputs to understand what's happening on each line.
+
+> ### {% icon question %} Question: Tracing Execution
+>
+> Create a table showing the numbers of the lines that are executed when this program runs,
+> and the values of the variables after each line is executed.
+>
+> ```
+> total = 0
+> for char in "tin":
+>     total = total + 1
+> ```
+> > ### {% icon solution %} Solution
+> >
+> > | Line | Variables            |
+> > |------|----------------------|
+> > | 1    | total = 0            |
+> > | 2    | total = 0 char = 't' |
+> > | 3    | total = 1 char = 't' |
+> > | 2    | total = 1 char = 'i' |
+> > | 3    | total = 2 char = 'i' |
+> > | 2    | total = 2 char = 'n' |
+> > | 3    | total = 3 char = 'n' |
+> {: .solution}
+{: .question}
+
+> ### {% icon question %} Question: Reversing a String
+>
+> Fill in the blanks in the program below so that it prints "stressed"
+> (the reverse of the original character string "desserts").
+>
+> ```
+> original = "stressed"
+> result = ____
+> for char in original:
+>     result = ____
+> print(result)
+> ```
+> > ### {% icon solution %} Solution
+> > ```
+> > original = "stressed"
+> > result = ""
+> > for char in original:
+> >     result = char + result
+> > print(result)
+> > ```
+> {: .solution}
+{: .question}
+
```

Comment:
```suggestion

```python
 #Test your code here !
original = "stressed"
result = ____
for char in original:
    result = ____
print(result)
```

```